rfc/20/README.md

```diff
@@ -0,0 +1,148 @@
+5. `$URL` is a placeholder (macro) that will be replaced with a real URL automatically by the external tool that execute the scripts from the yaml file.
+6. There is no guarantee that `$URL`s cohere between different components (i.e. component X can write key 'k' to `json:///tmp/my-file.json` while component Y tries to read that key from `yaml:///tmp/my-file.yaml`)
+7. Every stage must be idempotent. If the stage is applied twice in a row, the second run must succeed also. Effectively there must be no difference if the stage is applied once or N times in a row.
+8. Every stage is executed with root permissions.
+9. Scripts must use exit codes: 0 in case of SUCCESS, a number N > 0 if an ERROR occured.
+
+
+### Stage conditions
+
+| Stage        | Purpose           | Expected initial state  | Side effects performed as a result |
+| ------------- |-------------| ----- | --- |
+| post_install  | Verify that all the required 3rd party components are installed and have correct versions (Optional?) | Component RPM has just been installed to the OS. No services are started. | None (stage either fails or succeeds) |
+| config  | Apply changes to the OS to satisfy the needs of our component | Same to post_install | OS-level settings satisfy the needs of the component. |
+| init  | Initialize and start the component. | OS is ready to launch the component. | The component is started. |
+| test  | Sanity tests ran against the running component. The tests can produce some logs but they must not damage end user's data (reason: this stage can be executed in the field as well). | The component is started. | The component is started. The component looks working (no critical problems with the functionality were identified). |
+| reset  | Removes logs and other artifacts of test stage. | Makes no assumption on the component status. | Makes no assumption on the component status. |
+| cleanup  | Cleans the component-specific configuration. | TBD | The component is stopped, the configuration is cleared out. |
+| upgrade  | Perform the upgrade of the component (3rd-party components included). | TBD | TBD |
+| backup  | Save and export component's state so that it is state can be reproduced at another machine (TBD is it feasible?) | TBD | TBD |
+| restore  | An action opposite to `backup`. Must ingest the exported backup and apply it. | TBD | TBD |
+| support_bundle  | Prepare component-specific support bundle (TBD how will Provisioner Framework find all the files generated this way?) | TBD | TBD |
+
+
+## Communication with ConfStore
+
+TBD
+
+## Proposed changes to Hare
+
+1. All commands will be handled by `provisioning/setup.py` file.
+2. `provisioning/setup.py` must berenamed to `provisioning/setup_hare` (the file must be executable so proper shebang must be added to the first line).
+3. `setup_hare` must provide help screen.
+
+### Stage implementations
+
+#### post_install
+```
+setup_hare post_install --config <str>
+```
+Verifies that
+1. Consul is installed
```

Comment:
1. If I heard it correctly, consul will not be part of hare component. If it is correct, then this step may not be required, but I am not sure as hare has dependency on consul.
2. Is it required to verify these installations as well?
yum -y install python3 python3-devel
yum -y install yum-utils
dhall-bin
dhall-prelude

Reply:
Good question. I don't think I know the answer.

On one hand, our hare.spec file specifies those packages as dependencies already (except for yum-utils and python3-devel - those are required for developer environment only, dhall is installed and packages in our Makefile and thus doesn't need RPM):

```
Requires: consul = 1.7.8
Requires: facter
Requires: jq
Requires: cortx-motr = %{h_motr_version}
Requires: python36
```

So it doesn't make much sense to check that those RPMs are installed. Perhaps it is worth checking Consul version of the binary (we saw situations when there were multiple Consul versions were installed at LDR R1 cluster).

@mssawant, @NeeravChoudhari thoughts?